haskell/test/Test/ProtocolFee/XToY.hs

```diff
@@ -0,0 +1,893 @@
+        call cfmm (Call @"X_to_y") XToYParam
+          { xpDx = 53
+          , xpDeadline = deadline
+          , xpMinDy = 0
+          , xpToDy = swapper
+          }
+
+        -- sanity check
+        st <- getFullStorage cfmm
+        sCurTickIndex st @== -101
+
+        checkAllInvariants cfmm
+
+test_protocol_fees_are_burned :: TestTree
+test_protocol_fees_are_burned =
+  nettestScenarioOnEmulatorCaps "protocol fees are effectively burned" do
+    let feeBps = 0
+    let protoFeeBps = 8_000 -- 80%
+
+    liquidityProvider <- newAddress auto
+    swapper <- newAddress auto
+    let userTokenBalance = 1_e15
+    let accounts = [liquidityProvider, swapper]
+    let xTokenId = FA2.TokenId 0
+    let yTokenId = FA2.TokenId 0 -- doesn't actually matter
+    let xFa2storage = FA2.Storage
+          { sLedger = mkBigMap $ accounts <&> \acct -> ((acct, xTokenId), userTokenBalance)
+          , sOperators = mempty
+          , sTokenMetadata = mempty
+          }
+    ctezAdmin <- newAddress auto
+    let yCtezStorage = FA1_2.mkStorage ctezAdmin $
+          fromList $ accounts <&> (, userTokenBalance)
+    xToken <- originateSimple "fa2" xFa2storage (FA2.fa2Contract def { FA2.cAllowedTokenIds = [xTokenId] })
+    yToken <- originateSimple "ctez" yCtezStorage FA1_2.managedLedgerContract
+
+    cfmm <- originateSegCFMM FA2 CTEZ $ mkStorage xToken xTokenId yToken yTokenId feeBps protoFeeBps
+    checkAllInvariants cfmm
+
+    for_ accounts \account ->
+      withSender account do
+        call xToken (Call @"Update_operators") [FA2.AddOperator $ FA2.OperatorParam account (toAddress cfmm) xTokenId]
+        call yToken (Call @"Approve") (#spender .! (toAddress cfmm), #value .! userTokenBalance)
+
+    deadline <- mkDeadline
+    withSender liquidityProvider do
+      call cfmm (Call @"Set_position")
+        SetPositionParam
+          { sppLowerTickIndex = -100
+          , sppUpperTickIndex = 100
+          , sppLowerTickWitness = minTickIndex
+          , sppUpperTickWitness = minTickIndex
+          , sppLiquidity = 10_000
+          , sppDeadline = deadline
+          , sppMaximumTokensContributed = PerToken userTokenBalance userTokenBalance
+          }
+
+    -- Perform a swap that does not exhaust the position's liquidity
+    withSender swapper do
+      call cfmm (Call @"X_to_y") XToYParam
+        { xpDx = 10
+        , xpDeadline = deadline
+        , xpMinDy = 0
+        , xpToDy = swapper
+        }
+
+    -- The cfmm contract still has a CTEZ balance
+    cfmmBalance1 <- getFA12Balance yToken cfmm
+    checkCompares cfmmBalance1 (>=) 10
+
+    -- Trying to exhaust the position liquidity and consume most of the remaining
+    -- CTEZ balance is not possible (as it would without protocol fees)
+    withSender swapper do
+     call cfmm (Call @"X_to_y") XToYParam
+       { xpDx = 100
+       , xpDeadline = deadline
+       , xpMinDy = cfmmBalance1 - 10
+       , xpToDy = swapper
+       } & expectFailedWith smallerThanMinAssetErr
```

Comment:
I don't get this assertion. If the contract has `cfmmBalance1` tokens and ~10 are put aside due to protocol fees, why can't the user withdraw `cfmmBalance1 - 10` tokens?

Reply:
well for one thing, even if exactly 10 tokens were put aside for protocol fees by the previous swap, this new swap would also reserve some more, so they couldn't be given all to the user.

However truth be told I was using `10` as an arbitrary margin of error (the protocol fees here amount to `8` actually), which I see not that it's both confusing and doesn't serve as much of a proof.

I just pushed an update that should make this better, please take a look